### ISSUE TITLE
Latest version of conda takes too long to resolve dependencies

### ISSUE THREAD

**Issue by ocefpaf** [ioos/conda-recipes#473]
Latest version of conda abolished the `np` version string.  Even though that is a good thing, now we have a large number of possible combinations and the dependency solver is hanging up on travis!

One possible solution would be to delete the `nppy{27,34,35}` packages so the new ones (with only `py{27,34,35}`) can take their places.  We will lose download statistics, but that is not too bad.

@rsignell-usgs Are you OK with this solution?
@pelson Do you have a better solution in mind?

**Comment by emiliom** [comment]
@ocefpaf, does deleting the `nppy{27,34,35}` packages mean that one would no longer be able to export a requirements list from an older environments (created before this nppy/py switch), then use it on different machine to (re)create the same environment?

**Comment by ocefpaf** [comment]
That is hard to tell. If the export was done using a `conda env export` and the file has the build strings then the answer is, sadly, yes. :unamused: 

Conda's own documentation recommends to write the [environment.yml](https://github.com/ioos/secoora/blob/master/environment.yml) "by hand" to avoid that.

In theory, if only the version is referenced in the env file, the transition should be smooth.

IMHO conda's string-baking the `np{ver}` was toxic. That is why they removed it as the default behavior.  And later on they remove the empty `np` too.

Unfortunately, as early adopters of conda, we were left with thousands of binaries built that way (with `np18`, `np19`, `np`, and no `np`).  Those 4 options created a big matrix of dependency resolution that made conda choke.  In the best case scenario it solved after a long time.

This extra time represented a huge problem for the CIs.  We are limited to about ~30 to 40 minutes depending of the CI flavor.  I got hang-up from Travis-CI while conda tried to solve the dependency of just 1 package.  (See [this](https://travis-ci.org/ioos/conda-recipes/builds/83198417#L3139) matrix of 1000 possibilities!)

I know that conda devs are trying to fix that, but we could not afford to stay put until the problem goes away.  Also, the new releases without the `np` string are exactly the same as those with the `np` string.  So there is no need to provide both.  While those with the `np{ver}` should be considered the same as long as the package does not depend on the numpy ABI.  But I think we should specify that explicitly in the recipe file.

TL;DR

We have to encourage people to create [environment.yml](https://github.com/ioos/secoora/blob/master/environment.yml) files specifying only the version\*  and nothing else.  Letting conda solve the releases.

\* And even the version should only be specified when there is the need for complete reproducibility or a known issue with a certain version.  Otherwise follow the rule: new is always better :wink:

**Comment by emiliom** [comment]
YIKES! Thanks for the explanation. I'm glad you're on top of this.

> Conda's own documentation recommends to write the environment.yml "by hand" to avoid that.

Yikes, again!

> And even the version should only be specified when there is the need for complete reproducibility or a known issue with a certain version. Otherwise follow the rule: new is always better

I saw the exchange you, @rsignell-usgs and another USGS person had on this topic back in August or so, and thought about commenting. Yes, reproducibility and predictability (or managed changes) are pretty important when dealing with lots of near-real-time data streams and trying to maintain identical setups across machines. I'm working on this issue quite a bit this week and next. [Hence my offer to help test the merged, unversioned environment.yml across platforms](https://github.com/ioos/conda-recipes/pull/487#issuecomment-146684995).

**Comment by ocefpaf** [comment]
> I'm working on this issue quite a bit this week and next.

Thanks.  I just "solved" one [final issue](https://github.com/SciTools/conda-recipes-scitools/issues/115#issuecomment-146823225) with the env and I am merging #487

**Comment by pelson** [comment]
FWIW Looks like this is similar to https://github.com/conda/conda/issues/1616 and relates to https://github.com/conda/conda/issues/1338.

**Comment by ocefpaf** [comment]
Thank @pelson! I ended up removing the packages that had the `np{18,19}` and `np`.  The IOOS new builds do not bake the numpy string.  Hopefully will make due by just re-building the packages that break when a new version of numpy is release or by pinning the numpy version in the build/run session.

  That got our CIs back on business until https://github.com/conda/conda/issues/1338 is fixed.
